uobtheatre/bookings/models.py

```diff
@@ -521,36 +610,107 @@ def get_ticket_diff(
         )
 
     def pay(self, payment_method: "PaymentProvider") -> Optional["Transaction"]:
+        if self.is_reservation_expired:
+            raise CantBePaidForException(
+                message="This booking has expired. Please create a new booking"
+            )
+
+        # If this is a transfer and the booking it is transfered from is not paid
+        if self.transfered_from and self.transfered_from.status != Booking.Status.PAID:
+            self.transfered_from._check_transfer_booking()  # pylint: disable=protected-access
+
+        return super().pay(payment_method)
+
+    def complete(self, payment: Transaction = None):
+        """
+        Complete the booking (after it has been paid for) and send the
+        confirmation email.
         """
-        Pay for booking using provided payment method.
+        super().complete()
 
-        Args:
-            payment_method (PaymentMethod): The payment method used to pay for
-                the booking
+        # If the booking is transfered from another
+        if self.transfered_from:
+            self.transfered_from.status = Booking.Status.CANCELLED
+            self.transfered_from.save()
 
-        Returns:
-            Payment: The payment created by the checkout (optional)
+        self.send_confirmation_email(payment)
+
+    def clone(self):
+        clone = super().clone()
+        clone.reference = create_short_uuid()
+        return clone
+
+    def _check_transfer_performance(self, performance: "Performance"):
         """
-        # Cancel and delete pending payments for this booking
-        for payment in self.transactions.filter(status=Transaction.Status.PENDING):
-            payment.cancel()
+        Check that the performance can facilitate the transfer of this booking
+        to it
+        """
+        if self.performance == performance:
+            raise BookingTransferPerformanceUnchangedException
 
-        payment = payment_method.pay(self.total, self.misc_costs_value(), self)
+        if self.performance.production != performance.production:
+            raise BookingTransferToDifferentProductionException
 
-        # If a payment is created set the booking as paid
-        if payment.status == Transaction.Status.COMPLETED:
-            self.complete(payment)
+        if not performance.is_bookable:
+            raise NotBookableException
 
-        return payment
+        if performance.capacity_remaining < self.tickets.count():
+            raise CapacityException(
+                f"There are only {performance.capacity_remaining} seats remaining for the selected performance"
+            )
```

Comment:
This logic should be replaced with the ability once the PR with this code is merged: https://github.com/BristolSTA/uobtheatre-api/blob/35154eb132f084b076f0b0d0aec39429ac7e7ac3/uobtheatre/productions/abilities.py#L52